pkg/util/cache/intern.go

```diff
@@ -0,0 +1,359 @@
+}
+
+// KeyedInterner has an origin-keyed set of interners.
+type KeyedInterner struct {
+	interners      *lru.Cache
+	maxQuota       int
+	closeOnRelease bool
+	tmpPath        string
+	lastReport     time.Time
+	minFileSize    int64
+	maxPerInterner int
+	lock           sync.Mutex
+}
+
+// NewKeyedStringInterner creates a Keyed String Interner with a max per-origin quota of maxQuota
+func NewKeyedStringInterner(cfg cconfig.Component) *KeyedInterner {
+	stringInternerCacheSize := cfg.GetInt("dogstatsd_string_interner_size")
+	enableMMap := cfg.GetBool("dogstatsd_string_interner_mmap_enable")
+
+	if enableMMap {
+		closeOnRelease := !cfg.GetBool("dogstatsd_string_interner_mmap_preserve")
+		tempPath := cfg.GetString("dogstatsd_string_interner_tmpdir")
+		minSizeKb := cfg.GetInt("dogstatsd_string_interner_mmap_minsizekb")
+		maxStringsPerInterner := cfg.GetInt("dogstatsd_string_interner_origin_max_strings")
+		return NewKeyedStringInternerVals(stringInternerCacheSize, closeOnRelease, tempPath, minSizeKb, maxStringsPerInterner)
+	}
+	return NewKeyedStringInternerMemOnly(stringInternerCacheSize)
+}
+
+// NewKeyedStringInternerVals takes args explicitly for initialization
+func NewKeyedStringInternerVals(stringInternerCacheSize int, closeOnRelease bool, tempPath string, minFileKb, maxStringsPerInterner int) *KeyedInterner {
+	cache, err := lru.NewWithEvict(stringInternerCacheSize, func(_, internerUntyped interface{}) {
+		interner := internerUntyped.(*stringInterner)
+		interner.Release(1)
+	})
+	if err != nil {
+		return nil
+	}
+	return &KeyedInterner{
+		interners:      cache,
+		maxQuota:       -1,
+		lastReport:     time.Now(),
+		tmpPath:        tempPath,
+		minFileSize:    int64(minFileKb * 1024),
+		maxPerInterner: maxStringsPerInterner,
+		closeOnRelease: closeOnRelease,
+	}
+}
+
+// NewKeyedStringInternerMemOnly is a memory-only cache with no disk needs.
+func NewKeyedStringInternerMemOnly(stringInternerCacheSize int) *KeyedInterner {
+	cache, err := lru.NewWithEvict(stringInternerCacheSize, func(_, internerUntyped interface{}) {
+		interner := internerUntyped.(*stringInterner)
+		interner.Release(1)
+	})
+	if err != nil {
+		return nil
+	}
+	return &KeyedInterner{
+		interners:      cache,
+		maxQuota:       -1,
+		lastReport:     time.Now(),
+		tmpPath:        "",
+		maxPerInterner: initialInternerSize,
+		closeOnRelease: false,
+	}
+}
+
+// NewKeyedStringInternerForTest is a memory-only cache with a small default size.  Useful for
+// most tests.
+func NewKeyedStringInternerForTest() *KeyedInterner {
+	return NewKeyedStringInternerMemOnly(512)
+}
+
+var sGlobalQueryCount = atomic.NewInt64(0)
+var sFailedInternalCount = atomic.NewInt64(0)
+
+// LoadOrStoreString interns a string for an origin
+func (i *KeyedInterner) LoadOrStoreString(s string, origin string, retainer InternRetainer) string {
+	if Check(s) {
+		return i.LoadOrStore(unsafe.Slice(unsafe.StringData(s), len(s)), origin, retainer)
+	}
+	sFailedInternalCount.Add(1)
```

Comment:
Flagging another potential seqcst here. Although presumably rare if most strings are valid, pushing known invalid strings does ramp cost.

Reply:
The only way they fail the Check() is if there's a fairly severe bug.  I've been catching them with this function.  `sFailedInternalCount` is zero on my internal testing.  But I thought to leave the diagnostics in for any future changes that might need it.

I'm happy to take them out, or to hide behind a config flag.

Reply:
Nah, both are nit flags and this instance is less likely to fire than the other.